vis-cmds.c

```diff
@@ -632,16 +632,42 @@ static bool cmd_earlier_later(Vis *vis, Win *win, Command *cmd, const char *argv
 	return pos != EPOS;
 }
 
+static size_t space_replace(char *dest, const char *src, size_t maxlen) {
+	char *to = dest;
+	const char *from = src;
+	size_t destlen = 0;
+	size_t count = 0;
+	const size_t size = strlen("␣");
+	for (; *from != '\0' && destlen < maxlen - size - 1; from++) {
+		if (*from == ' ' && count < 3) {
+			strncpy(to, "␣", size);
+			count++;
+			destlen += size;
+			to += size;
+		} else {
+			*to = *from;
+			destlen++;
+			to++;
+		}
+	}
+	*to = '\0';
+	return count;
+}
+
 static bool print_keylayout(const char *key, void *value, void *data) {
-	return text_appendf(data, "  %-18s\t%s\n", key[0] == ' ' ? "␣" : key, (char*)value);
+	char buf[25];
+	size_t count = space_replace(buf, key, sizeof(buf));
+	return text_appendf(data, "  %-*s\t%s\n", 18+count*2, buf, (char*)value);
 }
 
 static bool print_keybinding(const char *key, void *value, void *data) {
 	KeyBinding *binding = value;
 	const char *desc = binding->alias;
 	if (!desc && binding->action)
 		desc = VIS_HELP_USE(binding->action->help);
-	return text_appendf(data, "  %-18s\t%s\n", key[0] == ' ' ? "␣" : key, desc ? desc : "");
```

Comment:
The `%-18s` here just means that the string length that will be printed
is 18 characters. If key is shorter than 18 chars it will be padded
with spaces. Otherwise 18 chars of key are printed and the rest is
truncated. It can just stay as is.

Reply:
Unfortunately, this does not work. With a constant padding of 18, I get misaligned text, for example:

![image](https://github.com/martanne/vis/assets/13970628/84d2fd6d-c306-4b22-a7b7-7553e554078a)

`[vsn]printf` with padding only counts bytes, not characters.

This can be verified with a minimal test:

```c
>> printf("%-4s|\n", " f");
 f  |
>> printf("%-4s|\n", "␣f");
␣f|
```

Reply:
I see, next time I should test before I comment.

In that case how about returning the corrected width. It should just be
`18 + count * size`. I think that would be better than adding the magic
number `2` in the code.

If you can make those corrections I think I will be fine with the patch.

Reply:
Sounds good to me. I'll change it like that.

Reply:
Please check my last commit. Now I return `invisiblebytecount`. The `2` was indeed magic. But the `18` would be magic, too, when located inside `space_replace`. And I still set `'\0'` (instead of `0`) for terminating the string because this is used in many places in this file.